language-plutus-core/src/Language/PlutusCore/Type.hs

```diff
@@ -91,29 +93,62 @@ instance Corecursive (Type tyname a) where
-                 => tyname a -- ^ @a@
-                 -> tyname a -- ^ @b@
-                 -> Type tyname a -- ^ @c@
-                 -> Type tyname a
-tyNameSubstitute tn tn' = cata a where
-    a (TyVarF x tn'') | tn == tn'' = TyVar x tn'
-    a x               = embed x
-
-instance (Eq (tyname a), Eq a) => Eq (Type tyname a) where
-    (==) (TyVar a1 tn) (TyVar a2 tn')                   = a1 == a2 && tn == tn'
-    (==) (TyFun a1 ty ty') (TyFun a2 ty'' ty''')        = a1 == a2 && ty == ty'' && ty' == ty'''
-    (==) (TyFix a1 tn ty) (TyFix a2 tn' ty')            =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty
-    (==) (TyForall a1 tn k ty) (TyForall a2 tn' k' ty') =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty && k == k'
-    (==) (TyBuiltin a1 b) (TyBuiltin a2 b')             = a1 == a2 && b == b'
-    (==) (TyInt a1 n) (TyInt a2 n')                     = a1 == a2 && n == n'
-    (==) (TyLam a1 tn k ty) (TyLam a2 tn' k' ty')       =
-        a1 == a2 && tyNameSubstitute tn' tn ty' == ty && k == k'
-    (==) (TyApp a1 ty ty') (TyApp a2 ty'' ty''')        = a1 == a2 && ty == ty'' && ty' == ty'''
-    (==) _ _                                            = False
+-- this type is used for replacing type names in
+-- the Eq instance
+type EqState tyname a = M.Map (tyname a) (tyname a)
+
+rebind :: (Ord (tyname a), MonadState (EqState tyname a) m)
+       => tyname a
+       -> tyname a
+       -> m ()
+rebind = modify .* M.insert
+
+remove :: (Ord (tyname a), MonadState (EqState tyname a) m)
+       => tyname a
+       -> m ()
+remove = modify . M.delete
+
+rebindAndEq :: (Eq a, Ord (tyname a), MonadState (EqState tyname a) m)
+            => Type tyname a
+            -> Type tyname a
+            -> tyname a
+            -> tyname a
+            -> m Bool
+rebindAndEq tyLeft tyRight tnLeft tnRight =
+    rebind tnRight tnLeft *>
+    eqTypeM tyLeft tyRight <* remove tnRight
+
+-- This tests for equality of names inside a monad that allows substitution.
+eqTypeM :: (Ord (tyname a), MonadState (EqState tyname a) m, Eq a)
+        => Type tyname a
+        -> Type tyname a
+        -> m Bool
+
+eqTypeM (TyFun _ ty ty') (TyFun _ ty'' ty''') = eqTypeM ty ty'' <&&> eqTypeM ty' ty'''
+eqTypeM (TyApp _ f a) (TyApp _ f' a') = eqTypeM f f' <&&> eqTypeM a a'
+
+eqTypeM (TyInt _ n) (TyInt _ n')              = pure (n == n')
+eqTypeM (TyBuiltin _ b) (TyBuiltin _ b')      = pure (b == b')
+
+eqTypeM (TyFix _ tnL tyL) (TyFix _ tnR tyR) =
+    rebindAndEq tyL tyR tnL tnR
+eqTypeM (TyForall _ tnL k tyL) (TyForall _ tnR k' tyR) = do
+    tyEq <- rebindAndEq tyL tyR tnL tnR
+    pure (k == k' && tyEq)
+eqTypeM (TyLam _ tnL k tyL) (TyLam _ tnR k' tyR) = do
+    tyEq <- rebindAndEq tyL tyR tnL tnR
+    pure (k == k' && tyEq)
+
+eqTypeM (TyVar _ tnRight) (TyVar _ tnLeft) = do
+    eqSt <- get
+    case M.lookup tnLeft eqSt of
+        Just tn -> pure (tnRight == tn)
+        Nothing -> pure (tnRight == tnLeft)
+
+eqTypeM _ _ = pure False
```

Comment:
You have four distinct naming conventions here:

```haskell
eqTypeM (TyFun _ ty ty') (TyFun _ ty'' ty''') = ...
eqTypeM (TyApp _ f a) (TyApp _ f' a') = ...
eqTypeM (TyFix _ tnL tyL) (TyFix _ tnR tyR) = ...
eqTypeM (TyVar _ tnRight) (TyVar _ tnLeft) = ...
```

That's confusing.

Reply:
Good point, I will fix this.